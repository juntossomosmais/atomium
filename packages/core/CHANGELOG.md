# Changelog

## 1.0.0 (2023-04-14)


### Bug Fixes

* use types from component class in stories file ([16f493c](https://github.com/juntossomosmais/atomium/commit/16f493c9b5d2017942a1eae26bc163fe1ee7458d))
