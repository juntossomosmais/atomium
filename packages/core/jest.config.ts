module.exports = {
  preset: '@stencil/core/testing',
  globals: {
    IONIC_CORE_ENTRY: 'node_modules/@ionic/core/dist/esm/index.js',
  },
  testEnvironment: 'jsdom',
  moduleNameMapper: {
    '^@ionic/core$': '<rootDir>/node_modules/@ionic/core/dist/esm/index.js',
  },
  transformIgnorePatterns: ['/node_modules/(?!@ionic|@stencil|@capacitor)'],
  setupFilesAfterEnv: ['@testing-library/jest-dom/extend-expect'],
  collectCoverage: true,
  coverageReporters: ['lcov'],
  coveragePathIgnorePatterns: ['/node_modules/'],
  collectCoverageFrom: [
    'src/components/**/*.{js,jsx,ts,tsx}',
    'output-target/**/*.{js,jsx,ts,tsx}',
    '!src/**/stories/**',
    '!src/**/*.mock.ts',
    '!src/**/*.spec.ts',
    '!output-target/**/*.spec.ts',
  ],
  coverageThreshold: {
    global: {
      lines: 80,
      statements: 80,
    },
  },
}
