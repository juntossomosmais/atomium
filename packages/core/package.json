{
  "name": "@juntossomosmais/atomium",
  "version": "3.12.2",
  "description": "Core of web components for Atomium",
  "repository": {
    "type": "git",
    "url": "https://github.com/juntossomosmais/atomium.git",
    "directory": "packages/core"
  },
  "publishConfig": {
    "@juntossomosmais:registry": "https://npm.pkg.github.com",
    "access": "public"
  },
  "main": "dist/index.cjs.js",
  "module": "dist/index.js",
  "es2015": "dist/esm/index.mjs",
  "es2017": "dist/esm/index.mjs",
  "types": "dist/types/index.d.ts",
  "collection": "dist/collection/collection-manifest.json",
  "collection:main": "dist/collection/index.js",
  "unpkg": "dist/core/core.esm.js",
  "files": [
    "dist/",
    "loader/",
    "react/",
    "vue/",
    "core.css"
  ],
  "scripts": {
    "build": "stencil build && npx copyfiles --flat ./dist/core/core.css ./",
    "start": "stencil build --watch",
    "test": "stencil test --spec",
    "test:ci": "stencil test --spec --coverage --no-cache --updateSnapshot --ci",
    "test:watch": "stencil test --spec",
    "generate": "stencil generate",
    "prebuild": "npx mkdirp react && npx mkdirp vue && cd > core.css",
    "publish-library": "npm publish --access public",
    "publish-library-alpha": "npm publish --access public --tag alpha"
  },
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/types/index.d.ts"
    },
    "./loader": {
      "types": "./loader/index.d.ts",
      "import": "./loader/index.js",
      "require": "./loader/index.cjs.js"
    },
    "./react": {
      "types": "./react/dist/types/index.d.ts",
      "import": "./react/dist/esm/index.js",
      "require": "./react/dist/cjs/index.js"
    },
    "./vue": {
      "types": "./vue/dist/types/index.d.ts",
      "import": "./vue/dist/esm/index.js"
    },
    "./core.css": {
      "import": "./core.css",
      "require": "./core.css"
    },
    "./dist/core/core.css": {
      "import": "./dist/core/core.css",
      "require": "./dist/core/core.css"
    }
  },
  "nx": {
    "targets": {
      "build": {
        "dependsOn": [
          "build",
          {
            "projects": [
              "@juntossomosmais/atomium-icons",
              "@juntossomosmais/atomium-tokens"
            ],
            "target": "build"
          }
        ]
      },
      "start": {
        "dependsOn": [
          "build"
        ]
      },
      "publish-library": {
        "dependsOn": [
          "build",
          {
            "projects": [
              "@juntossomosmais/atomium-react",
              "@juntossomosmais/atomium-vue"
            ],
            "target": "build"
          }
        ]
      }
    }
  }
}
