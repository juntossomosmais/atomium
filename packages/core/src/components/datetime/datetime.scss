.atom-datetime {
  &::part(calendar-day-active) {
    background-color: var(--color-brand-secondary-regular);
    color: var(--color-neutral-white);
  }

  &::part(calendar-day) {
    &:hover {
      background-color: var(--color-neutral-light-5);
    }
  }
}

.atom-item {
  position: relative;
}

.atom-label {
  color: var(--color-neutral-regular);
  font: var(--text-caption);
  left: 0;
  letter-spacing: var(--text-caption-letter);
  max-width: calc(100% / 0.75);
  overflow: hidden;
  padding-left: var(--spacing-base);
  position: absolute;
  text-overflow: ellipsis;
  top: 7px;
  white-space: nowrap;
  z-index: var(--zindex-5);
}

.atom-icon {
  color: var(--color-neutral-regular);
  pointer-events: none;
  position: absolute;
  right: var(--spacing-base);
  top: 50%;
  transform: translateY(-50%);
  z-index: var(--zindex-5);
}

.atom-button {
  border-bottom: 1px solid
    var(--ion-color-step-500, var(--ion-background-color-step-500, gray));
  width: 100%;

  .atom-item:hover & {
    &::part(native) {
      background: var(
        --ion-color-step-100,
        var(--ion-background-color-step-100, #e6e6e6)
      );
    }
  }

  &::part(native) {
    align-items: end;
    background: var(
      --ion-color-step-50,
      var(--ion-background-color-step-50, #f2f2f2)
    );
    border-end-end-radius: 0;
    border-end-start-radius: 0;
    border-start-end-radius: var(--spacing-base);
    border-start-start-radius: var(--spacing-base);
    color: var(--color-neutral-regular);
    display: flex;
    font: var(--text-body-medium);
    height: 57px;
    letter-spacing: var(--text-body-medium-letter);
    margin: 0;
    padding-bottom: 6px;
    padding-left: var(--spacing-base);
    padding-right: var(--spacing-base);
    text-align: left;
    width: 100%;
    z-index: var(--zindex-1);
  }

  &.date-active {
    border-bottom-color: var(--color-brand-secondary-regular);

    & ~ .atom-label {
      color: var(--color-brand-secondary-regular);
    }

    & ~ .atom-icon {
      color: var(--color-brand-secondary-regular);
    }
  }
}
