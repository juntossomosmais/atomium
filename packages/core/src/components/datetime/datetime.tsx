import {
  DatetimeChangeEventDetail,
  DatetimeCustomEvent,
  DatetimeHighlight,
  DatetimeHighlightCallback,
  DatetimePresentation,
} from '@ionic/core'
import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Prop,
  State,
  Watch,
} from '@stencil/core'

type TValue =
  | string
  | string[]
  | DatetimeCustomEvent
  | DatetimeChangeEventDetail
  | undefined

@Component({
  tag: 'atom-datetime',
  styleUrl: 'datetime.scss',
  scoped: true,
})
export class AtomDatetime {
  @Element() element!: HTMLElement

  @Prop() label?: string
  @Prop() cancelText?: string
  @Prop() clearText?: string
  @Prop() datetimeId = 'datetime'
  @Prop() dayValues?: number[] | string
  @Prop() disabled?: boolean
  @Prop() doneText?: string
  @Prop() formatOptions: {
    date?: Intl.DateTimeFormatOptions
    time?: Intl.DateTimeFormatOptions
  } = {
    date: {
      month: '2-digit',
      day: '2-digit',
      year: 'numeric',
    },
    time: {
      hour: '2-digit',
      minute: '2-digit',
    },
  }
  @Prop() highlightedDates?: DatetimeHighlight[] | DatetimeHighlightCallback
  @Prop() hourCycle: 'h12' | 'h23' = 'h23'
  @Prop() hourValues?: number[] | string
  @Prop() locale = 'pt-BR'
  @Prop() isDateEnabled?: (dateString: string) => boolean
  @Prop() max?: string
  @Prop() min?: string
  @Prop() minuteValues?: number[] | string
  @Prop() monthValues?: number[] | string
  @Prop() multiple?: boolean
  @Prop() name?: string
  @Prop() preferWheel = false
  @Prop() presentation?: DatetimePresentation = 'date'
  @Prop() rangeMode?: boolean
  @Prop() readonly?: boolean
  @Prop() showClearButton?: boolean
  @Prop() showDefaultButtons?: boolean
  @Prop() showDefaultTimeLabel = false
  @Prop() showDefaultTitle = false
  @Prop() useButton = false
  @Prop() size?: 'cover' | 'fixed' = 'fixed'
  @Prop({ mutable: true, reflect: true }) value?: TValue

  @Prop() yearValues?: number[] | string

  @State() selectedDates: string[] = []
  @State() private ionDatetimeValue: string[] | undefined = undefined

  @Event() atomFocus!: EventEmitter<void>
  @Event() atomBlur!: EventEmitter<void>
  @Event() atomChange!: EventEmitter<string | string[] | undefined>
  @Event() atomCancel!: EventEmitter<void>

  private _datetimeEl!: HTMLIonDatetimeElement

  get datetimeEl(): HTMLIonDatetimeElement {
    return this._datetimeEl
  }

  set datetimeEl(value: HTMLIonDatetimeElement) {
    this._datetimeEl = value
  }

  private filterEmptyStrings(arr: string[]): string[] | undefined {
    const filteredArray = arr.filter((item) => item?.trim() !== '')

    return filteredArray.length > 0 ? filteredArray : undefined
  }

  private handleStringValue(val: string): string[] | undefined {
    return val.trim() !== '' ? [val] : undefined
  }

  private handleCustomEventValue(
    val: DatetimeCustomEvent
  ): string[] | undefined {
    const detailValue = val.detail?.value

    if (typeof detailValue === 'string') {
      return this.handleStringValue(detailValue)
    }

    if (Array.isArray(detailValue)) {
      return this.filterEmptyStrings(detailValue)
    }

    return undefined
  }

  private normalizeValue(val: TValue): string[] | undefined {
    if (!val) {
      return undefined
    }

    if (Array.isArray(val)) {
      return this.filterEmptyStrings(val)
    }

    if (typeof val === 'string') {
      return this.handleStringValue(val)
    }

    if ('detail' in val && val.detail?.value !== undefined) {
      return this.handleCustomEventValue(val as DatetimeCustomEvent)
    }

    return undefined
  }

  componentWillLoad() {
    const normalizedValue = this.normalizeValue(this.value)

    if (this.rangeMode && normalizedValue) {
      this.selectedDates = this.calculateDateRange(normalizedValue)
    } else if (normalizedValue) {
      this.selectedDates = normalizedValue
    } else {
      this.selectedDates = []
    }

    // WORKAROUND: Ionic's ion-datetime can sometimes reset to a default year (e.g., 2021)
    // if its 'value' prop is set too early during component initialization or state updates.
    // Setting the value with a setTimeout(..., 0) allows the ion-datetime to fully render
    // before its value is programmatically applied, preventing this bug.
    setTimeout(() => {
      let valueToSet: string[] | undefined

      if (this.rangeMode) {
        valueToSet = [...this.selectedDates]
      } else {
        valueToSet =
          this.selectedDates.length > 0 ? [this.selectedDates[0]] : undefined
      }

      this.ionDatetimeValue = valueToSet
    }, 0)
  }

  @Watch('value')
  onValueChange(newValue: TValue) {
    const normalizedValue = this.normalizeValue(newValue)

    if (this.rangeMode && normalizedValue) {
      this.selectedDates = this.calculateDateRange(normalizedValue)
    } else if (normalizedValue) {
      this.selectedDates = normalizedValue
    } else {
      this.selectedDates = []
    }

    // WORKAROUND: See explanation in componentWillLoad. Applied on prop changes too.
    setTimeout(() => {
      let valueToSet: string[] | undefined

      if (this.rangeMode) {
        valueToSet = [...this.selectedDates]
      } else {
        valueToSet =
          this.selectedDates.length > 0 ? [this.selectedDates[0]] : undefined
      }

      this.ionDatetimeValue = valueToSet
    }, 0)
  }

  private handleRangeMode(dates: string[]) {
    const sortedDates = [...dates].sort()

    if (sortedDates.length === 2) {
      this.selectedDates = this.calculateDateRange(sortedDates)
    } else if (sortedDates.length === 1) {
      this.selectedDates =
        this.selectedDates.length >= 2
          ? sortedDates
          : [...this.selectedDates, ...sortedDates]
    } else {
      this.selectedDates = []
    }

    this.selectedDates = [...this.selectedDates]
    this.atomChange.emit(this.selectedDates)

    // WORKAROUND: See explanation in componentWillLoad. Applied after user range selection.
    setTimeout(() => {
      this.ionDatetimeValue = [...this.selectedDates]
    }, 0)
  }

  handleDateChange = (event: CustomEvent<DatetimeChangeEventDetail>) => {
    const rawValue = event.detail.value
    let dates: string[]

    if (Array.isArray(rawValue)) {
      dates = rawValue
    } else if (!rawValue) {
      dates = []
    } else {
      dates = [rawValue]
    }

    if (this.rangeMode) {
      this.handleRangeMode(dates)

      return
    }

    // Fix: handle multiple mode (not range)
    if (this.multiple && !this.rangeMode) {
      this.selectedDates = dates
      this.atomChange.emit(this.selectedDates)

      // WORKAROUND: See explanation in componentWillLoad. Applied on prop changes too.
      setTimeout(() => {
        this.ionDatetimeValue = [...this.selectedDates]
      }, 0)

      return
    }

    const singleValue = dates.length > 0 ? dates[0] : undefined

    this.atomChange.emit(singleValue)

    // WORKAROUND: See explanation in componentWillLoad. Applied on prop changes too.
    setTimeout(() => {
      this.ionDatetimeValue = singleValue ? [singleValue] : undefined
    }, 0)
  }

  private readonly handleBlur = () => {
    this.atomBlur.emit()
  }

  private readonly handleFocus = () => {
    this.atomFocus.emit()
  }

  private readonly handleCancel = () => {
    this.atomCancel.emit()
  }

  private calculateDateRange(dates: string[]): string[] {
    if (dates.length < 2) return dates

    let [start, end] = dates.map((dateStr) => this.parseLocalDate(dateStr))

    if (start.getTime() > end.getTime()) {
      [start, end] = [end, start]
    }

    const dayMilliseconds = 1000 * 60 * 60 * 24
    const daysDifference = Math.round(
      Math.abs((end.getTime() - start.getTime()) / dayMilliseconds)
    )

    const dateRange: string[] = []

    for (let i = 0; i <= daysDifference; i++) {
      const currentDate = new Date(start)

      currentDate.setDate(start.getDate() + i)
      dateRange.push(this.formatDateToISOString(currentDate))
    }

    return dateRange
  }

  private parseLocalDate(dateString: string): Date {
    const [year, month, day] = dateString.split('-').map(Number)

    return new Date(year, month - 1, day)
  }

  private formatDateToISOString(date: Date): string {
    const year = date.getFullYear()
    const month = (date.getMonth() + 1).toString().padStart(2, '0')
    const day = date.getDate().toString().padStart(2, '0')

    return `${year}-${month}-${day}`
  }

  private getRangeLabel(): string | null {
    if (!this.rangeMode || this.selectedDates.length < 2) return null

    const sortedSelectedDates = [...this.selectedDates].sort(
      (a, b) => new Date(a).getTime() - new Date(b).getTime()
    )
    const [start, end] = [
      sortedSelectedDates[0],
      sortedSelectedDates[sortedSelectedDates.length - 1],
    ]

    const startDate = this.parseLocalDate(start)
    const endDate = this.parseLocalDate(end)
    const format = (date: Date) =>
      date.toLocaleDateString(this.locale || 'pt-BR', {
        day: '2-digit',
        month: '2-digit',
        year: 'numeric',
      })

    return `${format(startDate)} - ${format(endDate)}`
  }

  private getDateTargetSlot(): React.ReactNode {
    if (!this.rangeMode) return <slot name='date-target' />

    if (this.selectedDates.length === 0) {
      return <span slot='date-target'>Selecionar data</span>
    }

    if (this.selectedDates.length === 1) {
      const date = this.parseLocalDate(this.selectedDates[0])
      const formatted = date.toLocaleDateString(this.locale || 'pt-BR', {
        day: '2-digit',
        month: '2-digit',
        year: 'numeric',
      })

      return <span slot='date-target'>{formatted}</span>
    }

    return (
      <span slot='date-target' class='custom-date-target'>
        {this.getRangeLabel()}
      </span>
    )
  }

  private renderDatetime() {
    return (
      <ion-datetime
        ref={(el) => (this.datetimeEl = el as HTMLIonDatetimeElement)}
        class='atom-datetime'
        color='secondary'
        cancelText={this.cancelText}
        clearText={this.clearText}
        dayValues={this.dayValues}
        disabled={this.disabled}
        doneText={this.doneText}
        formatOptions={this.formatOptions}
        highlightedDates={this.highlightedDates}
        hourCycle={this.hourCycle}
        hourValues={this.hourValues}
        id={this.datetimeId}
        isDateEnabled={this.isDateEnabled}
        locale={this.locale}
        max={this.max?.trim() || undefined}
        min={this.min?.trim() || undefined}
        minuteValues={this.minuteValues}
        monthValues={this.monthValues}
        multiple={this.multiple || this.rangeMode}
        mode='md'
        name={this.name}
        presentation={this.presentation}
        preferWheel={this.preferWheel}
        readonly={this.readonly}
        showClearButton={this.showClearButton}
        showDefaultButtons={this.showDefaultButtons}
        showDefaultTimeLabel={this.showDefaultTimeLabel}
        showDefaultTitle={this.showDefaultTitle}
        size={this.size}
        yearValues={this.yearValues}
        value={
          this.multiple || this.rangeMode
            ? this.ionDatetimeValue
            : (this.ionDatetimeValue?.[0] ?? undefined)
        }
        onIonChange={this.handleDateChange}
        onIonCancel={this.handleCancel}
        onIonBlur={this.handleBlur}
        onIonFocus={this.handleFocus}
      >
        {this.showDefaultTimeLabel && (
          <span slot='time-label'>
            <slot name='time-label' />
          </span>
        )}
        {this.showDefaultTitle && (
          <span slot='title'>
            <slot name='title' />
          </span>
        )}
      </ion-datetime>
    )
  }

  private renderWithButton(): React.ReactNode {
    return (
      <div>
        <div class='atom-item'>
          <ion-datetime-button
            class='atom-button'
            color='secondary'
            datetime={this.datetimeId}
            disabled={this.disabled}
            mode='md'
          >
            {this.getDateTargetSlot()}
            <slot name='time-target' />
          </ion-datetime-button>
          <span class='atom-label'>{this.label}</span>
          <atom-icon
            class='atom-icon'
            icon={
              this.presentation === 'time'
                ? 'clock-outline'
                : 'calendar-today-outline'
            }
          />
        </div>
        <ion-popover
          keep-contents-mounted='true'
          show-backdrop='false'
          style={{ 'margin-top': '1px' }}
        >
          {this.renderDatetime()}
        </ion-popover>
      </div>
    )
  }

  render() {
    if (this.useButton) {
      return <Host>{this.renderWithButton()}</Host>
    } else {
      return <Host>{this.renderDatetime()}</Host>
    }
  }
}
