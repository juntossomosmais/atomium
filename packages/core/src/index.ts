export { AtomButton } from './components/button/button'
export { AtomCol, AtomGrid, AtomRow } from './components/grid'
export { AtomIcon } from './components/icon/icon'
export { AtomInput } from './components/input/input'
export { AtomSelect } from './components/select/select'
