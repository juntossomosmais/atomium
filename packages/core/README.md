# Atomium - Web Components

This is the Web Components implementation of the Atomium design system to be used in any Framework or Vanilla JS

## Getting Started

### Installation

```bash
npm i @juntossomosmais/atomium
```

### Basic usage example

```jsx
import '@juntossomosmais/atomium/core.css'

import { defineCustomElements } from '@juntossomosmais/atomium/loader'

defineCustomElements(window)

<atom-button>Click me</atom-button>
```

#### Using with CDN

```html
<head>
  <link rel="stylesheet" href="https://atomium.juntossomosmais.com.br/core/SEMVER/core.css">
  <link rel="stylesheet" href="https://atomium.juntossomosmais.com.br/tokens/SEMVER/tokens.css">
</head>
<body>
  <atom-button>Click me</atom-button>

  <script type="module">
    import { defineCustomElements } from 'https://atomium.juntossomosmais.com.br/core/SEMVER/loader/index.js'

    document.addEventListener('DOMContentLoaded', function () {
      defineCustomElements()
    })
  </script>
</body>
```

