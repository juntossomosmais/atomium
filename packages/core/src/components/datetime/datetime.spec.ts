import { newSpecPage } from '@stencil/core/testing'

import { AtomDatetime } from './datetime'

describe('AtomDatetime', () => {
  it('renders with all props', async () => {
    const labelText = 'Select a date'

    const page = await newSpecPage({
      components: [AtomDatetime],
      html: `<atom-datetime label="${labelText}"></atom-datetime>`,
    })

    await page.waitForChanges()

    expect(page.root).toEqualHtml(`
      <atom-datetime label="${labelText}">
        <ion-datetime class="atom-datetime" color="secondary" hourcycle="h23" id="datetime" locale="pt-BR" mode="md" presentation="date" size="fixed"></ion-datetime>
      </atom-datetime>
    `)
  })

  it('should renders with a button when useButton is true', async () => {
    const page = await newSpecPage({
      components: [AtomDatetime],
      html: '<atom-datetime useButton="true"></atom-datetime>',
    })

    await page.waitForChanges()

    expect(page.root).toEqualHtml(`
      <atom-datetime usebutton="true">
        <ion-datetime class="atom-datetime" color="secondary" hourcycle="h23" id="datetime" locale="pt-BR" mode="md" presentation="date" size="fixed"></ion-datetime>
      </atom-datetime>
    `)
  })

  it('should emits atomChange with multiple dates when rangeMode is true', async () => {
    const page = await newSpecPage({
      components: [AtomDatetime],
      html: '<atom-datetime range-mode="true"></atom-datetime>',
    })

    await page.waitForChanges()

    const datetime = page.root?.querySelector('ion-datetime')
    const spy = jest.fn()

    page.root?.addEventListener('atomChange', spy)

    datetime?.dispatchEvent(
      new CustomEvent('ionChange', {
        detail: { value: ['2022-01-01', '2022-01-02'] },
      })
    )

    expect(spy).toHaveBeenCalled()
    expect(spy.mock.calls[0][0].detail).toEqual(['2022-01-01', '2022-01-02'])
  })

  it('should start a new range when a new date is selected after a range', async () => {
    const page = await newSpecPage({
      components: [AtomDatetime],
      html: '<atom-datetime range-mode="true"></atom-datetime>',
    })

    const datetime = page.rootInstance
    const spy = jest.fn()

    page.root?.addEventListener('atomChange', spy)

    datetime.handleRangeMode(['2022-01-01', '2022-01-05'])
    await page.waitForChanges()

    datetime.handleRangeMode(['2022-01-10'])
    await page.waitForChanges()

    expect(spy).toHaveBeenCalledTimes(2)
    expect(spy.mock.calls[1][0].detail).toEqual([
      '2022-01-01',
      '2022-01-02',
      '2022-01-03',
      '2022-01-04',
      '2022-01-05',
      '2022-01-10',
    ])
  })

  it('should emits atomChange with a single date when rangeMode is false', async () => {
    const page = await newSpecPage({
      components: [AtomDatetime],
      html: '<atom-datetime range-mode="false"></atom-datetime>',
    })

    await page.waitForChanges()

    const datetime = page.root?.querySelector('ion-datetime')
    const spy = jest.fn()

    page.root?.addEventListener('atomChange', spy)

    datetime?.dispatchEvent(
      new CustomEvent('ionChange', {
        detail: { value: '2022-01-01' },
      })
    )

    expect(spy).toHaveBeenCalled()
    expect(spy.mock.calls[0][0].detail).toEqual('2022-01-01')
  })

  it('should emits atomCancel when cancel button is clicked', async () => {
    const page = await newSpecPage({
      components: [AtomDatetime],
      html: '<atom-datetime></atom-datetime>',
    })

    await page.waitForChanges()

    const datetime = page.root?.querySelector('ion-datetime')
    const spy = jest.fn()

    page.root?.addEventListener('atomCancel', spy)

    datetime?.dispatchEvent(new CustomEvent('ionCancel'))

    expect(spy).toHaveBeenCalled()
  })

  it('should emits atomFocus when datetime is focused', async () => {
    const page = await newSpecPage({
      components: [AtomDatetime],
      html: '<atom-datetime></atom-datetime>',
    })

    await page.waitForChanges()

    const datetime = page.root?.querySelector('ion-datetime')
    const spy = jest.fn()

    page.root?.addEventListener('atomFocus', spy)

    datetime?.dispatchEvent(new CustomEvent('ionFocus'))

    expect(spy).toHaveBeenCalled()
  })

  it('should emits atomBlur when datetime is blurred', async () => {
    const page = await newSpecPage({
      components: [AtomDatetime],
      html: '<atom-datetime></atom-datetime>',
    })

    await page.waitForChanges()

    const datetime = page.root?.querySelector('ion-datetime')
    const spy = jest.fn()

    page.root?.addEventListener('atomBlur', spy)

    datetime?.dispatchEvent(new CustomEvent('ionBlur'))

    expect(spy).toHaveBeenCalled()
  })

  it('should render custom dateTarget when dateTarget slot is used', async () => {
    const page = await newSpecPage({
      components: [AtomDatetime],
      html: `<atom-datetime>
        <div slot="date-target">Custom Date Target</div>
      </atom-datetime>`,
    })

    await page.waitForChanges()

    expect(page.root).toEqualHtml(`
      <atom-datetime>
        <div slot="date-target">
          Custom Date Target
        </div>
        <ion-datetime class="atom-datetime" color="secondary" hourcycle="h23" id="datetime" locale="pt-BR" mode="md" presentation="date" size="fixed"></ion-datetime>
      </atom-datetime>
    `)
  })
})
