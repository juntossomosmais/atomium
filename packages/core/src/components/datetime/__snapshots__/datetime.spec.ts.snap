// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtomDatetime renders with all props 1`] = `
<atom-datetime label="Select a date">
  <!---->
  <ion-datetime class="atom-datetime" color="secondary" hourcycle="h23" id="datetime" locale="pt-BR" mode="md" presentation="date" size="fixed"></ion-datetime>
</atom-datetime>
`;

exports[`AtomDatetime should renders with a button when useButton is true 1`] = `
<atom-datetime usebutton="true">
  <!---->
  <ion-datetime class="atom-datetime" color="secondary" hourcycle="h23" id="datetime" locale="pt-BR" mode="md" presentation="date" size="fixed"></ion-datetime>
</atom-datetime>
`;
